### PR TITLE
feat: Add `vector` based `to_awkward` API

### LINKED ISSUE #21: awkward-array API
would be nice to be able to read this with awkward-array. Perhaps using some of the "behavioral interfaces"  we've been discussing with @jpivarski

### PR DESCRIPTION
**Pull request by lukasheinrich**
resolves #21 

this is a first stab at scikit-hep integration and making pylhe nicely integrated w/ `awkward` and `vector`

cc @jpivarski @henryiii @eduardo-rodrigues @matthewfeickert 

eventually we might drop the old non-awkward classes

image

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Add awkward and vector as core dependencies
* Add API to build awkward arrays from pylhe events
* Add example notebook using the awkward API
* Add tests for output of dir(pylhe.awkward)

Co-authored-by: Matthew Feickert <matthew.feickert@cern.ch>
```

### CODE REVIEW

**Comment by eduardo-rodrigues** [comment]
Morning @lukasheinrich, sounds like a great idea to me. I just wonder if you want to have the dependency a requirement or rather do it optionally in case one wants the awkward backend, as done in other packages for "extra" dependencies (as you do int for good reasons with several backends in `pyhf`, for example).

It may be good to keep the non-awkward way as that's simple and maybe expected by many (a theorist for example).

**Comment by lukasheinrich** [comment]
thoughts @matthewfeickert ? Having an `pylhe[awkward]` extra sounds good to me but I also would be willing  to take it on as a fixed dep

**Comment by matthewfeickert** [comment]
> thoughts @matthewfeickert ? Having an `pylhe[awkward]` extra sounds good to me but I also would be willing to take it on as a fixed dep

I would be okay with an extra, but a dependency seems pretty reasonable to me given that `awkward` just brings itself and `numpy` (c.f. below). I personally think that trying to move toward using `awkward` is a good idea from an ecosystem view as it would help teach `awkward` literacy to people &mdash; and give improved performance! :)

```
root@820bbf6e56a4:/# pip install pylhe
root@820bbf6e56a4:/# pip list
Package    Version
---------- -------
attrs      20.3.0
decorator  4.4.2
hepunits   2.1.0
networkx   2.5.1
particle   0.14.1
pip        21.0.1
pylhe      0.2.1
setuptools 56.0.0
tex2pix    0.3.1
wheel      0.36.2
root@820bbf6e56a4:/# pip install awkward
root@820bbf6e56a4:/# pip list
Package    Version
---------- -------
attrs      20.3.0
awkward    1.2.2
decorator  4.4.2
hepunits   2.1.0
networkx   2.5.1
numpy      1.20.2
particle   0.14.1
pip        21.0.1
pylhe      0.2.1
setuptools 56.0.0
tex2pix    0.3.1
wheel      0.36.2
```

File: src/pylhe/__init__.py [hunk 1]
```diff
@@ -7,6 +7,8 @@
 import tex2pix
 from particle.converters.bimap import DirectionalMaps
 
+from .awkward import register_awkward, to_akward
```
Comment (matthewfeickert) [inline src/pylhe/__init__.py:10]: Failing flake8 here

```
./src/pylhe/awkward.py:1:1: F401 'numpy as np' imported but unused
./src/pylhe/awkward.py:35:32: E713 test for membership should be 'not in'
./src/pylhe/__init__.py:10:1: F401 '.awkward.register_awkward' imported but unused
./src/pylhe/__init__.py:10:1: F401 '.awkward.to_akward' imported but unused
```

File: src/pylhe/awkward.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+__all__ = ["register_awkward", "to_awkward"]
+
+
+# Python 3.7+
+def __dir__():
+    return __all__
+
+
+def register_awkward():
```
Comment (matthewfeickert) [inline src/pylhe/awkward.py:12]: Would be good to get some docstrings here, but not necessary to merge this now.

Comment (matthewfeickert) [inline src/pylhe/awkward.py:12]: Would be good to get some docstrings here, but not necessary to merge this now.

File: src/pylhe/awkward.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    vector.register_awkward()
+    ak.mixin_class(ak.behavior)(Particle)
+    ak.mixin_class(ak.behavior)(Event)
+    ak.mixin_class(ak.behavior)(EventInfo)
+
+
+def to_awkward(event_iterable):
```
Comment (matthewfeickert) [inline src/pylhe/awkward.py:19]: Would be good to get some docstrings here, but not necessary to merge this now.

**Review by matthewfeickert — APPROVED** [review]
@lukasheinrich I did some revisions to get this passing pre-commit and also passing most of CI, but this LGTM overall now. :+1: 

I've also rebased this, so feel free to undo any of my revisions, but just tag me if you have any questions. I think this is good to go for now.

**Pull request merged**